### PR TITLE
Apple WWDC 2020 logo variant

### PR DESCRIPTION
**Pull request by twang817**
## Description
Adds Apple's WWDC 2020 logo as a variant.

## Types of changes :
- Bug fix (non-breaking change which fixes an issue)
- Typo / style fix (non-breaking change which improves readability)
- [x] New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to change)

## Checklist :
- \[IF NEEDED\] I have updated the _README.md_ file accordingly ;
- \[IF NEEDED\] I have updated the test cases (which pass) accordingly ;
- \[IF BREAKING\] This pull request targets next Archey version branch ;
- [x] My changes looks good ;
- [x] I agree that my code may be modified in the future ;
- [x] My code follows the code style of this project ([PEP8](https://www.python.org/dev/peps/pep-0008/)).

image

### CODE REVIEW

File: archey/logos/darwin.py
Comment (HorlogeSkynet) [inline archey/logos/darwin.py:55]: Hi @twang817 ! Thanks for your PR.

Some thoughts :

* your ASCII art is **really** big (despite being pretty though 😇) : it weighs around 21KB, which is disproportionate according to others. Do you think you could rework it to get closer to the "classic" dimensions ?

File: archey/output.py
Comment (HorlogeSkynet) [inline archey/output.py:48]: I don't think we need this `preformat` flag and "options" notion here : if your logo don't contain any `c[ ]` values, it will be formatted "as-is" 👍

File: archey/logos/darwin.py
Comment (twang817) [reply archey/logos/darwin.py:55]: I didn't draw this. I came across the wwdc logo and copied it from there. 

https://www.reddit.com/r/apple/comments/4ob7q2/anyone_else_love_the_asciistyle_apple_logo_from/?utm_source=share&utm_medium=ios_app&utm_name=iossmf

I wouldn't know where to begin with shrinking this.

File: archey/output.py
Comment (twang817) [reply archey/output.py:48]: This issue is with the .format(), since the string contains unmatched {}.

**Comment by ingrinder** [comment]
> I didn't draw this. I came across the wwdc logo and copied it from there.

Are you sure we can add this logo then? I'm not 100% sure on the intricacies of licensing here, but I would surmise that we can't relicense the work made by Apple... If it were recreated in the same style, I would suppose that might count as an "original" work.

**Comment by twang817** [comment]
> > I didn't draw this. I came across the wwdc logo and copied it from there.
> 
> Are you sure we can add this logo then? I'm not 100% sure on the intricacies of licensing here, but I would surmise that we can't relicense the work made by Apple... If it were recreated in the same style, I would suppose that might count as an "original" work.

I honestly didn't give it anywhere near this much thought. The picture isn't exactly the same. I used a random generator for the internal symbols and randomized the colors. 

That being said, if there are issues, I would wonder what the legality of any of the logos/trademarks are. I suspect that there isn't anyone that's going to get upset with this use case. 

If the size is a concern, I'd happily share my Python code that generates the ascii file. It's stupid code, done in a hurry. Perhaps it could be worked into something that generates and caches the logo on first load.

As for shrinking it, I really don't have a single artistic bone in my body. Perhaps someone else will come along and redo this in a smaller size.

Really, I made this change for my own use. I've always liked this graphic, and I thought I'd share. I'm not trying to sound stand-offish, but at this point, it's entirely up to you.

**Comment by HorlogeSkynet** [comment]
I'll try to work on this within some days :+1:

**Comment by HorlogeSkynet** [comment]
So I caught a glimpse of it, me points below :

* License _should_ not be an issue here (partly-related : dylanaraps/neofetch#170).
  But that raised a question though, under which license the Reddit OP released this ? :sweat_smile:
* This logo is definitely **too** big ;
* The Reddit thread had been started 6 years ago : is this really the WWDC **2020** logo ?
* If there are some unmatched `{`, please prefer escaping them (`{{`) :wink:
* Would your script be able to output something like ~20/30 lines high ? "Resolution" wouldn't look clean as yours but maybe it can be a good trade-off ?

Thanks again, bye :pray:

**Comment by twang817** [comment]
I've uploaded the code I used to generate this.  I wrote this some time ago, and revisited it today.

A couple points:
1. It uses ply.  A bit crazy, but I must have been in a lex-yacc mood that day.
2. It generates from an existing logo.  It figures out the edges to retain shape and then figures out how to follow a word/non-word pattern and randomize the symbols, words, and colors.
3. It does have a stretch built in that can stretch it horizontally.  I think I used this to accommodate for fonts that may be skinnier than they are tall.

I don't know how useful this is to your requirements and/or goals.

**Comment by HorlogeSkynet** [comment]
Hey @twang817, sorry for the delay.

So I've checked your gist, see below my points/suggestions :

* The logo height _seems_ OK (it is pretty "tall" according to others, but it is not very important)
* Do you think you can (with your script of course) _colorize_ tokens using the `{c[N]}` syntax (literally 6 bytes in the end source code) instead of explicit control characters (i.e. `\033[...`) full format ?
* Of course, you can provision a (Python) list `COLORS_WWDC` containing all the colors that are needed.
* If it's still "too big", maybe we can limit coloration to a subset of tokens (words and specific characters tokens) ?

Thanks again, bye :wave:

**Comment by ingrinder** [comment]
I noticed this has been around for a while so I've had a bit of a go at addressing some of the above concerns.

I forked the gist with a few small changes to get it to output an Archey-style logo and dirtily remove redundant colour codes - https://gist.github.com/ingrinder/9e68f7bd1e2d8c9d86a0a5a594a64e1d

I also made some tentative changes to allow the specific colours to be used by adding 256-colour support, see the logo/apple_wwdc20 branch.

image

It's still "big", at 8K and 36 lines, but it looks fairly decent to me despite the size? I suppose the big empty space satisfies the Apple aesthetic...! I wouldn't really know how to start removing lines from the logo without redrawing the whole input one by hand. @HorlogeSkynet if you think this is viable, perhaps I could write some tests and @twang817 could pull it into this PR?

**Comment by HorlogeSkynet** [comment]
Thanks @ingrinder for the hard work !!! I'm ready to review a PR for your branch 🙏

**Comment by ingrinder** [comment]
@twang817 I pushed onto the end of your fork, I hope that's okay - it seemed easier than making a separate PR, and I believe you won't lose credit in the commit once it's merged that way.

@HorlogeSkynet I think this is ready for review, I added tests and squashed together some of my commits. Quick overview of changes:

- Splits `Colors` class into `Style` base class for general formatting stuff (possibility for extension in future with other attributes)
- Adds `Colors8Bit` class extending `Style` to allow 256 colour mode text to be used (as in this PR)
- Adds modified WWDC20 logo in "Archey style" from gist linked above
- Unit tests for new classes

Let me know what you think :+1:

File: archey/colors.py
Comment (HorlogeSkynet) [inline archey/colors.py:27]: Hey @ingrinder !
Many thanks for the hard work, it's a very clean patch.
My two cents below :

1. I wonder when/why this case occurs ? Do we _really_ need to catch this exception ? 🙄
2. I think we should directly amend @twang817's first commit here, at least for these reasons : it would prevent a huge revision introducing 21+KB additions (and another one removing them just after) from reaching project history (as well as your 153da7f90830a633a5b596774e175f65dc52511a revert could directly "fix up" the first one here). I guess a first commit with your `Style` implementation and another one adding the WWDC 2020 logo afterwards would be optimal. What do you think ?

Thanks again ! Bye 👋

Comment (ingrinder) [reply archey/colors.py:27]: I suppose it's not really necessary to catch, but if someone somewhere makes a `Style` and then tries `str` on it, it will raise `AttributeError` unless its value has been explicitly set (which should be ideally from a subclass).
I thought about giving it an `__init__` which sets `value`, but this also causes an `AttributeError` using it as a mixin with an `Enum` class since `value` is immutable in that case.

I suppose we could remove this, probably should be making an instance of `Style` in the first place, although maybe we could write an `__init__(self, *value)` which either checks `isinstance(self, Enum)` or catches an `AttributeError` before setting `value` (EAFP), then it's required to be defined if an instance is made:

```python
    def __init__(self, *value):
        try:
            self.value = value
        except AttributeError:
            # can't set value (probably an `Enum`)
            pass

    def __str__(self):
        if self.should_color_output():
            return self.escape_code_from_attrs(";".join(map(str, self.value)))
        return ""
```

Though admittedly even that feels a little bit hacky :frowning_face:

As for 2, I was trying to avoid force-pushing onto the fork :smile: but let me know which approach you think is best and I'll do an interactive rebase to squash it all into two commits as you suggest (I wasn't sure if you just wanted to squash and merge in one!). :wave:

Comment (HorlogeSkynet) [reply archey/colors.py:27]: 1. Wow, what a mess 👀
    My opinion is that we shouldn't catch it in the first place, if one day someone uses `Style` directly and then calls `str` it will fail and they'll need to fix their code anyway.
    On the other hand, if one day we drop support for "old" Python releases, we could update this class to a more appropriate object (if there is).
3. I'm okay for a squash in one or two commits (preference for two as I suggested, this way we "uncouple" your `Colors` rework/extension from the WWDC logo actual addition) 😉

Many thanks again, cheers 👋

**Comment by ingrinder** [comment]
@HorlogeSkynet Rebased on master and squashed into 2 commits - I removed the `AttributeError` check as you said. Should be good to merge if you're happy :tada:

**Review by HorlogeSkynet — APPROVED** [review]

**Comment by HorlogeSkynet** [comment]
Many thanks @ingrinder :pray: 
@twang817 This will be released in the next Archey version, optimistically before the end of July.

**Pull request merged**
